src/librustc/mir/mod.rs

```diff
@@ -2474,14 +2474,34 @@ impl<'tcx> UserTypeProjections<'tcx> {
     }
 }
 
-#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash, RustcEncodable, RustcDecodable)]
+#[derive(Clone, Debug, PartialEq, Eq, Hash, RustcEncodable, RustcDecodable)]
 pub struct UserTypeProjection<'tcx> {
```

Comment:
I think we should definitely add a comment to this type explaining what-the-heck is going on. Here is my attempt:

---

Encodes a user-supplied type annotation -- the type annotation is determined by applying the given list of projections to an underlying base type. Often, the projection-element list is empty, in which case this simply encodes a type. But in the case of complex patterns we may need to apply only a "subset" of the type to a variable, in which case the `projs` vector is used.

Example:

- `let x: T = ...` -- here, the `projs` vector is empty
- `let (x, _): T = ...` -- here, the `projs` vector would contain the `0` field, indicating that the type of `x` is determined by finding the type of the `0` field from `T`

Reply:
Oy you are correct, I had a "todo" for myself in a commit message to add docs and then I think I either ignored it, or deleted it during a rebase

Reply:
added adaptation of above text to the docs for `UserTypeProjection`.